### PR TITLE
Update 1-click trial process to go via Markeplace e-shop

### PR DESCRIPTION
**Pull request by michalkleiner**
### Description:

Ref DEV-18349.

The behaviour on narrower screens was tweaked in dev, that may need adjusting, though I'm not sure what other options are there.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

File: plugins/Marketplace/Marketplace.php [hunk 1]
```diff
@@ -167,7 +167,6 @@ public function getClientSideTranslationKeys(&$translationKeys)
         $translationKeys[] = 'Marketplace_AutoUpdateDisabledWarning';
         $translationKeys[] = 'Marketplace_ByXDevelopers';
         $translationKeys[] = 'Marketplace_ClickToCompletePurchase';
-        $translationKeys[] = 'Marketplace_CurrentNumPiwikUsers';
```
Comment (michalkleiner) [inline plugins/Marketplace/Marketplace.php:170]: This line was duplicated, so removed one instance of it.

File: plugins/Marketplace/stylesheets/plugin-details.less [hunk 2]
```diff
@@ -455,9 +459,15 @@
         padding: 24px;
         height: 90px;
         border-top: 1px solid #aaa;
+        margin-top: 1px; // to prevent images overflowing the border
```
Comment (michalkleiner) [inline plugins/Marketplace/stylesheets/plugin-details.less:468]: Unrelated to the change, but I notices some of the floating images were overlapping the border, so this fixes it.

File: plugins/Marketplace/lang/en.json [hunk 3]
```diff
@@ -11,6 +11,7 @@
         "TryFreeTrialTitle": "Try 30 days for free, then",
         "Free": "Free",
         "FreeTrialLabel" : "Free Trial",
+        "Free30DayTrialThen": "Free 30-day trial, then",
```
Comment (sgiehl) [inline plugins/Marketplace/lang/en.json]: Any reason for not reusing the existing `TryFreeTrialTitle`. The text might be slightly different, but already has translations to various languages.
If we stick to this new one, we should remove `TryFreeTrialTitle` from the translation file.

**Review by sgiehl — CHANGES_REQUESTED** [review]
Generally seems to work as expected. Left a minor translation related comment.

Besides that the UI tests for starting a trial are now failing. Those need to be removed and/or adjusted to cover the new flow.

And I came across one thing that seems weird. For a plugin, which is installed, but the license is missing the super user sees a license missing warning, which is correct. But he normal user has a request trial button, while seeing the no license warning in the details modal.  It will most likely not be possible at all to start a new trial in that case, as if the plugin is already installed, there most likely had been a trial or purchase before, so maybe we should simply not show the trial button in that case.

image

File: plugins/Marketplace/lang/en.json (hunk 3 above)
Comment (michalkleiner) [reply plugins/Marketplace/lang/en.json]: The new title is a middle ground between what it was before and what Javi had in his designs in a way where it reads as a sentence with the dropdown and makes it clear that it's free first AND after the trial you pay XYZ euros or dollars, so basically accessibility reasons.
I can revert it to the old one, or remove the old one and keep new one 🤷 
I get the point with the translations though.

Comment (michalkleiner) [reply plugins/Marketplace/lang/en.json]: Used the old one for the benefit of the existing translations.

**Comment by michalkleiner** [comment]
> And I came across one thing that seems weird. For a plugin, which is installed, but the license is missing the super user sees a license missing warning, which is correct. But he normal user has a request trial button, while seeing the no license warning in the details modal. It will most likely not be possible at all to start a new trial in that case, as if the plugin is already installed, there most likely had been a trial or purchase before, so maybe we should simply not show the trial button in that case.
> 
@sgiehl what combination of user access level did you use to get the screenshot? I can't reproduce it. When the plugin is installed and license is removed, a super user gets the notification that there's no license, and a regular user gets "More details" (after a tiny change), but a regular user never sees the notification as it requires "hasSomeAdminAccess" to be true.

**Comment by michalkleiner** [comment]
> Besides that the UI tests for starting a trial are now failing. Those need to be removed and/or adjusted to cover the new flow.

@sgiehl removed the Start Free Trial UI tests for now.

**Comment by michalkleiner** [comment]
> And I came across one thing that seems weird. For a plugin, which is installed, but the license is missing the super user sees a license missing warning, which is correct. But he normal user has a request trial button, while seeing the no license warning in the details modal. It will most likely not be possible at all to start a new trial in that case, as if the plugin is already installed, there most likely had been a trial or purchase before, so maybe we should simply not show the trial button in that case.

@sgiehl I've confirmed this was happening for a user that had admin access to another site ("some admin access"). After a change they see 'More details' on the plugin card, no license notification in the modal, but no 'Request trial' within the modal.

**Review by sgiehl — APPROVED** [review]
The previously reported things all seem fixed.
But I came across something else, but as that most likely was already the case before, it might not need to be fixed here.

When running Matomo on PHP 7.2, without a license for premium plugins, as super user it shows the `More Details` button for `LoginSaml`. I guess the `Start Trial` isn't shown, as it requires PHP 7.3. But the modal shows this:

image

So it shows that it can't be installed, but the warning with the reason is missing.

**Comment by michalkleiner** [comment]
I've checked how this behaved on 5.0.0 before the on-prem marketplace changes and the warning wasn't shown either for premium plugins. I'll create an internal ticket to follow that up.

**Pull request merged**